### PR TITLE
Spsa optimization

### PR DESCRIPTION
**Pull request by elrama-**
Implemented the SPSA algorithm for the adaptive control of MC

Implemented the tests for this algorithm

For future: After acquiring more experience, decide on optimal values for the gain parameters

### CODE REVIEW

**Comment by AdriaanRol** [comment]
@elrama- Picture? (e.g., the 2 parameter example)

**Comment by elrama-** [comment]
image

both colors and gray (previous run) are working examples, the difference is in the gain parameters of the algorithm (the ones that would be nice to know more about to do nice initial guesses).

The test implemented in tests_MC is the color line.

File: pycqed/measurement/optimization.py [hunk 1]
```diff
@@ -135,3 +135,99 @@ def nelder_mead(fun, x0,
...
+            and operate over a numpy array of the same dimensions as x0
+        x0 (numpy array): initial position
+
+
+        no_improv_thr,  no_improv_break (float, int): break after
+            no_improv_break iterations with an improvement lower than
+            no_improv_thr
+        maxiter (int): always break after this number of iterations.
+            Set it to 0 to loop indefinitely.
+        alpha, gamma, a, c, A: parameters for the SPSA gains
```
Comment (AdriaanRol) [inline pycqed/measurement/optimization.py]: @elrama- could you add the expected type ('m assuming float) to the docstring of these args? (better to be consistent with the rest.

Comment (elrama-) [inline pycqed/measurement/optimization.py]: done

**Comment by AdriaanRol** [comment]
@elrama- I was about to merge this but noticed you added some extra parameters that I don't know what they do and you commented out some piece of code. Could you add what they do to the docstring (or refer the source of the docs as in the rest?) then it would be absolutely perfect!

**Pull request merged**